design/Fancy calling conventions-design.md

```diff
@@ -0,0 +1,118 @@
+# CSE231 Final Project Design
+
+Members:
+
+- Lien-Bee Huang (lih004@ucsd.edu)
+- Xi-Zhen Liu (xil087@ucsd.edu)
+- Yang-Hang Wu (yaw038@ucsd.edu)
+
+## Interfaces
+
+### AST
+
+```typescript=
+
+export type Parameter<A> = { 
+    name: string, 
+    type: Type,
+    value: Expr<A>,
+}
+export type FunDef<A> = { 
+    a?: A, 
+    name: string, 
+    parameters: Array<Parameter<A>>, 
+    ret: Type, 
+    inits: Array<VarInit<A>>, 
+    body: Array<Stmt<A>>,
+    arbarg_idx: number,
+    kwarg_idx: number,
+}
+```
+
+To deal with parameters with default value, we would like to add an attribute call `value` in `Parameter<A>` to store the default value. It could be any expression at the parsing stage. If there's no specific default value, the attribute will be set to `null`. This can also help us distinguish `None` from no default value.
+To deal with arbitrary arguments and keyword arguments, we would like to add attribute `arbarg_idx` to store the index of the tuple for the arbitrary arguments, and `kwarg_idx` to store the index of the dictionary for the keyword arguments. Since we will see all the arbitrary arguments as a tuple, and all the keyword arguments as a dictionary, we store the indices of them to distinguish them from normal tuple and dict parameters.
```

Comment:
Doesn't `**kwargs` have to go last?

Reply:
Also this allows representing `*args` and `**kwargs` with default values, which doesn't seem great.

Could we instead just have two different fields for them? Is their position relevant?

Reply:
At first, yes, the position matters.
For example, the 2 functions below would be different
```python=
def f(a, *args):
    print(a)
f(1,2,3)
```

```python=
def f(*args, a):
    print(a)
f(1,2,3)
```
Also the name of arbitrary arguments could be other name from `args` or `kwargs`. (maybe `*a`)
To support that, treating them as normal arguments would be more reasonable than using 2 different fields.
And for the default values part, we can do some extra check to avoid this problem.

For the position of `**kwargs`, yes, it need to be the last one.
But to check if arbitrary keyword arguments allowed and maintain the similar style with `*args`, we decide to use index to keep the information